### PR TITLE
增加DubboBootstrap启动停止事件扩展

### PR DESCRIPTION
**Pull request by theonefx**
增加 DubboBootstrapStartStopListener 来对外通知 DubboBootstarp 的启动停止回调扩展；
向Spring环境抛出DubboBootstarp 的启动停止事件。

#7150

### CODE REVIEW

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/bootstrap/DubboBootstrapStartStopListener.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package org.apache.dubbo.config.bootstrap;
+
+import org.apache.dubbo.common.extension.SPI;
+
+/**
+ * call on DubboBootstrap start or stop.
+ *
+ * @author <a href="mailto:chenxilzx1@gmail.com">theonefx</a>
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/bootstrap/DubboBootstrapStartStopListener.java:25]: Hi, @theonefx,
 
Please make sure no @author tag gets appended to the file you contribute to as the @author tag is incompatible with Apache. Rest assured, other ways, including CVS, will ensure transparency, fairness in recording your contributions.

You can mention this at https://github.com/apache/dubbo/blob/master/CONTRIBUTING.md#code-conventions.

Comment (theonefx) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/bootstrap/DubboBootstrapStartStopListener.java]: done

**Review by AlbumenJ — APPROVED** [review]
LGTM.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapStartStopListenerSpringAdapter.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.context.ApplicationContext;
+
+/**
+ * convcert Dubbo bootstrap event to spring environment.
+ *
+ * @scene 2.7.9
+ */
+public class DubboBootstrapStartStopListenerSpringAdapter implements DubboBootstrapStartStopListener {
+
+    static ApplicationContext applicationContext;
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapStartStopListenerSpringAdapter.java:33]: It’s not good to use a static field here:
``` java
static ApplicationContext applicationContext;
```

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapApplicationListener.java [hunk 3]
```diff
@@ -46,6 +46,7 @@
     public DubboBootstrapApplicationListener(ApplicationContext applicationContext) {
         super(applicationContext);
         this.dubboBootstrap = DubboBootstrap.getInstance();
+        DubboBootstrapStartStopListenerSpringAdapter.applicationContext = applicationContext;
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapApplicationListener.java:49]: Is there a better way to initialize `DubboBootstrapStartStopListenerSpringAdapter`?

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapStartStopListenerSpringAdapter.java (hunk 2 above)
Comment (theonefx) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/context/DubboBootstrapStartStopListenerSpringAdapter.java:33]: two reason:
1. The DubboBootstrapStartStopListenerSpringAdapter instance generated by Dubbo and invoked by Dubbo. so i can't set any property to it.
2. The "instance" field of DubboBootstrap is static. So using the same dimensions (ie static variables) is a reasonable solution.

**Pull request merged**